modin/data_management/functions/groupby_function.py

```diff
@@ -20,174 +20,243 @@
 class GroupbyReduceFunction(MapReduceFunction):
     @classmethod
     def call(cls, map_func, reduce_func, *call_args, **call_kwds):
-        def caller(
-            query_compiler,
-            by,
-            axis,
-            groupby_args,
-            map_args,
-            reduce_args=None,
-            numeric_only=True,
-            drop=False,
-        ):
-            if not isinstance(by, (type(query_compiler), str)):
-                by = try_cast_to_pandas(by, squeeze=True)
-                return query_compiler.default_to_pandas(
-                    lambda df: map_func(
-                        df.groupby(by=by, axis=axis, **groupby_args), **map_args
-                    )
+        assert not (
+            isinstance(map_func, dict) ^ isinstance(reduce_func, dict)
+        ) and not (
+            callable(map_func) ^ callable(reduce_func)
+        ), "Map and reduce functions must be either both dict or both callable."
+
+        return lambda *args, **kwargs: cls.caller(
+            *args, map_func=map_func, reduce_func=reduce_func, **kwargs, **call_kwds
+        )
+
+    @classmethod
+    def map(
+        cls,
+        df,
+        other=None,
+        axis=0,
+        by=None,
+        groupby_args=None,
+        map_func=None,
+        map_args=None,
+        drop=False,
+    ):
+        # Set `as_index` to True to track the metadata of the grouping object
+        # It is used to make sure that between phases we are constructing the
+        # right index and placing columns in the correct order.
+        groupby_args["as_index"] = True
+        groupby_args["observed"] = True
+        if other is not None:
+            other = other.squeeze(axis=axis ^ 1)
+            if isinstance(other, pandas.DataFrame):
+                df = pandas.concat(
+                    [df] + [other[[o for o in other if o not in df]]],
+                    axis=1,
                 )
-            assert axis == 0, "Can only groupby reduce with axis=0"
+                other = list(other.columns)
+            by_part = other
+        else:
+            by_part = by
+
+        apply_func = cls.try_filter_dict(map_func, df)
```

Comment:
It is necessary to leave the columns that are in the partition?

Reply:
yes, otherwise we will get a key error
```python
>>> df
   col0  col1  col2  col3
0     0     0     0     0
1     1     1     1     1
2     2     2     2     2
3     3     3     3     3
>>> df.groupby("col0").agg({"col1": "sum", "missing_column": "prod"})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\site-packages\pandas\core\groupby\generic.py", line 951, in aggregate
    result, how = self._aggregate(func, *args, **kwargs)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python37_64\lib\site-packages\pandas\core\base.py", line 354, in _aggregate
    raise SpecificationError(f"Column(s) {cols} do not exist")
pandas.core.base.SpecificationError: Column(s) ['missing_column'] do not exist
```